src/bin/cargo-fixit/fixit.rs

```diff
@@ -50,54 +51,59 @@ fn exec(args: FixitArgs) -> CargoResult<()> {
         .ok()
         .and_then(|i| i.parse().ok())
         .unwrap_or(4);
+    let mut iteration = 0;
 
-    let mut last_errors = IndexSet::new();
-    let mut last_made_changes = false;
+    let mut last_errors;
 
-    for _ in 0..max_iterations {
-        (last_errors, last_made_changes) = run_rustfix(&args, &mut files)?;
+    let mut current_package_id = None;
+    let mut seen = HashSet::new();
 
-        if !last_made_changes {
+    loop {
+        let (errors, fixed) = run_rustfix(&args, &mut files, &mut current_package_id, &seen)?;
+
+        last_errors = errors;
+        if let Some(p) = fixed {
+            trace!("fixed `{}` in {} iteration(s)", p, iteration);
+            seen.insert(p);
+            iteration = 0;
+        }
+
+        iteration += 1;
+
+        if let Some(p) = &current_package_id {
+            if iteration >= max_iterations {
+                trace!("max iterations {} reached for `{}`", iteration, p);
+                seen.insert(p.to_owned());
+                current_package_id = None;
+                iteration = 0;
+            }
+        } else {
             break;
         }
     }
     for (name, file) in files {
         shell::fixed(name, file.fixes)?;
     }
-
-    if last_made_changes {
-        let only = HashSet::new();
-        let output = std::process::Command::new(env!("CARGO"))
-            .args(["check", "--message-format", "json"])
-            .stderr(Stdio::piped())
-            .stdout(Stdio::piped())
-            .output()?;
-        for line in String::from_utf8(output.stdout)?.lines() {
-            let diagnostic = match serde_json::from_str::<CheckMessage>(line) {
-                Ok(check) => check.message,
-                _ => continue,
-            };
-            if collect_suggestions(&diagnostic, &only, rustfix::Filter::Everything).is_some() {
-                if let Some(rendered) = diagnostic.rendered {
-                    eprint!("{}\n\n", rendered.trim_end());
-                }
-            };
-        }
-    } else {
-        for e in last_errors {
-            eprint!("{}\n\n", e.trim_end());
-        }
+    for e in last_errors {
+        eprint!("{}\n\n", e.trim_end());
     }
 
     Ok(())
 }
 
+/// Returns `Ok((errors, Some(package_id)))` when `package_id` is fixed completely
+///
+/// Returns `Ok((errors, None))` when the `current_package_id` is not fixed completely
 fn run_rustfix(
```

Comment:
In the first pass, lets let the caller decide whether things are fully fixed, like it currently does